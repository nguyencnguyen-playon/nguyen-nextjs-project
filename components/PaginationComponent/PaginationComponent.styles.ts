import { css } from 'styled-components';

export const styles = {
    wrapper: css`
    display: flex;
    flex-direction: column;
    border: 1px solid #dbdbdb;
    border-left-width: 0;
    border-right-width: 0;
    border-bottom-width: 0;
    padding: 32px;
    margin-top: 16px;
  `,
    buttonsSection: css<{ align: string }>`
    display: flex;
    flex-direction: row;
    justify-content: ${({ align }) => align};
    align-items: center;
  `,
    dropdownSection: css`
    display: flex;
    flex-direction: row;
    flex: 1;
    align-items: center;
    height: 100%;
  `,
    listItemsContainer: css`
    display: flex;
    flex-direction: row;
    justify-content: 'flex-start';
    width: 100%;
    flex-wrap: wrap;
  `,
    pageNumberButtonsContainer: css<{ margin: string }>`
    display: flex;
    flex-direction: row;
  `,
    pageNumberButton: css<{ active: boolean; btnBackgroundColor: string }>`
    align-items: center;
    justify-content: center;
    width: 30px;
    height: 40px;
    border: 1px solid #dbdbdb;
    background-color: ${({ active, btnBackgroundColor }) => (active ? btnBackgroundColor : 'white')};
  `,
    pageNumberText: css<{ active: boolean; color: string }>`
    font-weight: 700;
    color: ${({ active, color }) => (active ? color : '#000')};
  `,
    navigatePageButton: css`
    align-items: center;
    justify-content: center;
    width: 40px;
    height: 40px;
    border: 1px solid #dbdbdb;
    background-color: white;
  `,
    navigatePageText: css<{ disabled: boolean }>`
    font-weight: 700;
    color: ${({ disabled }) => (disabled ? '#dbdbdb' : '#000')};
  `
};
