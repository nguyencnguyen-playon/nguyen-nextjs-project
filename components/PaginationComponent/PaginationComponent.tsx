import styled from 'styled-components';
import React from 'react';
import { Dropdown } from './DropDown';
import { PageNumberButtons } from './PageNumberButtons';
import { styles } from './PaginationComponent.styles';

const Wrapper = styled.div`
  ${styles.wrapper}
`;

const ButtonsSection = styled.div<{ align: string }>`
  ${styles.buttonsSection}
`;

const DropdownSection = styled.div`
  ${styles.dropdownSection}
`;

const ListItemsContainer = styled.div`
  ${styles.listItemsContainer}
`;

export const PageRanges = (currentPage = 0, totalPages = 1) => {
    if (totalPages <= 3) {
        return [1, totalPages];
    }

    if (currentPage === 1) {
        return [1, 3];
    }

    // If we are in the last page, show the last 3 pages
    if (currentPage === totalPages) {
        return [totalPages - 2, totalPages];
    }

    return [currentPage - 1, currentPage + 1];
};

type PaginationComponentProps = {
    currentPage: number;
    totalPages?: number;
    pageSize?: number;
    data: any[];
    onChange: (newPage: number) => void;
    onPageSizeChange: (newPageSize: string) => void;
    renderItems: (items: any, index: number) => React.ReactNode;
    isPaginationEnabled?: boolean;
    btnBackgroundColor?: string;
    textBtnColor?: string;
};

export const PaginationComponent = ({
    currentPage,
    totalPages = 1,
    pageSize = 10,
    data,
    onChange,
    onPageSizeChange,
    renderItems,
    isPaginationEnabled,
    btnBackgroundColor = '#ffffff',
    textBtnColor = '#000000'
}: PaginationComponentProps) => {
    const handlePageChange = (newPage: number) => {
        onChange(newPage);
    };

    const [startPage, endPage] = PageRanges(currentPage, totalPages);
    /* Render page number buttons */

    return !isPaginationEnabled ? (
        <ListItemsContainer data-testid='pagination-component-disabled'>
            {data.map((item, index) => renderItems(item, index))}
        </ListItemsContainer>
    ) : (
        <Wrapper data-testid='pagination-component'>
            <ButtonsSection align='space-between'>
                <DropdownSection>
                    <Dropdown
                        onValueChange={pageSize => onPageSizeChange(pageSize)}
                        selectedValue={pageSize?.toString() || '5'}
                        options={[5, 10, 20].map(opt => ({
                            label: opt.toString() || '',
                            value: opt.toString() || ''
                        }))}
                    ></Dropdown>
                </DropdownSection>
                <PageNumberButtons
                    margin='0 0 10px 0'
                    currentPage={currentPage}
                    totalPages={totalPages}
                    handlePageChange={handlePageChange}
                    textBtnColor={textBtnColor}
                    btnBackgroundColor={btnBackgroundColor}
                    endPage={endPage}
                    startPage={startPage}
                />
            </ButtonsSection>
            <ListItemsContainer>
                {data.map((item, index) => renderItems(item, index))}
            </ListItemsContainer>
            <ButtonsSection align='flex-end'>
                <PageNumberButtons
                    margin='10px 0 0 0'
                    currentPage={currentPage}
                    totalPages={totalPages}
                    handlePageChange={handlePageChange}
                    textBtnColor={textBtnColor}
                    btnBackgroundColor={btnBackgroundColor}
                    endPage={endPage}
                    startPage={startPage}
                />
            </ButtonsSection>
        </Wrapper>
    );
};
