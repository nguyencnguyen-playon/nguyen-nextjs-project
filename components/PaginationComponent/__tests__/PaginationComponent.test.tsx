import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { PageRanges, PaginationComponent } from '../PaginationComponent';
describe('PaginationComponent', () => {
    const mockData = [
        {
            "id": 1715015431309,
            "desc": "45454",
            "name": "5454545",
            "image_url": "https://picsum.photos/992"
        },
        {
            "id": 1715179628607,
            "desc": "web8",
            "name": "web8",
            "image_url": "https://picsum.photos/459"
        }
    ];
    const mockOnChange = jest.fn();
    const mockOnPageSizeChange = jest.fn();
    const mockRenderItems = jest.fn();

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders correctly with default props', () => {
        const component = render(
            <PaginationComponent
                currentPage={1}
                data={mockData}
                onChange={mockOnChange}
                onPageSizeChange={mockOnPageSizeChange}
                renderItems={mockRenderItems}
                totalPages={2}
                isPaginationEnabled
            />
        );

        // Add your assertions here to check if the component renders correctly
        // For example:
        expect(component).toMatchSnapshot();
    });

    it('calls onChange when page number is clicked', () => {
        const { getAllByText } = render(
            <PaginationComponent
                currentPage={1}
                data={mockData}
                onChange={mockOnChange}
                onPageSizeChange={mockOnPageSizeChange}
                renderItems={mockRenderItems}
                totalPages={2}
                isPaginationEnabled
            />
        );
        const elements = getAllByText('1');
        elements.forEach(element => {
            fireEvent.click(element)
        });

        expect(mockOnChange).toHaveBeenCalledWith(1);
    });

    it('should not render pagination layout with isPaginationEnabled = false', () => {
        const { getByTestId } = render(
            <PaginationComponent
                currentPage={1}
                data={mockData}
                onChange={mockOnChange}
                onPageSizeChange={mockOnPageSizeChange}
                renderItems={mockRenderItems}
                totalPages={2}
                isPaginationEnabled={false}
            />
        );

        expect(getByTestId('pagination-component-disabled')).toBeTruthy();
    })

    it('should render pagination layout with isPaginationEnabled = true', () => {
        const { getByTestId } = render(
            <PaginationComponent
                currentPage={1}
                data={mockData}
                onChange={mockOnChange}
                onPageSizeChange={mockOnPageSizeChange}
                renderItems={mockRenderItems}
                isPaginationEnabled
            />
        );

        expect(getByTestId('pagination-component')).toBeTruthy();
    })

    it('result PageRanges return should be correctly', () => {
        const result = PageRanges(1, 10);
        expect(result).toEqual([1, 3]);

    })
    it('should return last 3 pages when currentPage is the last page', () => {
        const result = PageRanges(10, 10);
        expect(result).toEqual([8, 10]);
    })
    it('should call handlePageChange when page number is clicked', () => {
        const { getAllByText } = render(
            <PaginationComponent
                currentPage={1}
                data={mockData}
                onChange={mockOnChange}
                onPageSizeChange={mockOnPageSizeChange}
                renderItems={mockRenderItems}
                isPaginationEnabled
                totalPages={2}
            />
        );
        const elements = getAllByText('1');
        elements.forEach(element => {
            fireEvent.click(element)
        });

        expect(mockOnChange).toHaveBeenCalledWith(1);
    });
});
