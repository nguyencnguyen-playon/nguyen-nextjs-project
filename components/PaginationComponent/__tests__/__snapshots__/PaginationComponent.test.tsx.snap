// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationComponent renders correctly with default props 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-qRuZU gmucBJ"
        data-testid="pagination-component"
      >
        <div
          align="space-between"
          class="sc-jsMaUZ dKTrZm"
        >
          <div
            class="sc-kFuvnu iOKxMI"
          >
            <div>
              Show 
            </div>
            <div
              class="sc-beqWNU gbqjiq"
            >
              <select
                class="sc-gueXAH fwRQqY"
              >
                <option
                  class="sc-dmqGRC"
                  value="5"
                >
                  5
                </option>
                <option
                  class="sc-dmqGRC"
                  value="10"
                >
                  10
                </option>
                <option
                  class="sc-dmqGRC"
                  value="20"
                >
                  20
                </option>
              </select>
              <div
                class="sc-hLseSp eewNXK"
              >
                &gt;
              </div>
            </div>
          </div>
          <div
            class="sc-eDDOjk hzWDLl"
            margin="0 0 10px 0"
          >
            <button
              class="sc-iAEzLF gRRRHm"
              disabled=""
            >
              <div
                class="sc-fsQjcA iKmgdh"
                disabled=""
              >
                &lt;
              </div>
            </button>
            <button
              btnbackgroundcolor="#ffffff"
              class="sc-jTrOVX kwSTpv"
            >
              <div
                class="sc-gLDymS jVaaBL"
                color="#000000"
              >
                1
              </div>
            </button>
            <button
              btnbackgroundcolor="#ffffff"
              class="sc-jTrOVX cLKghP"
            >
              <div
                class="sc-gLDymS TLaLT"
                color="#000000"
              >
                2
              </div>
            </button>
            <button
              class="sc-iAEzLF gRRRHm"
            >
              <div
                class="sc-fsQjcA hFwLtP"
              >
                &gt;
              </div>
            </button>
          </div>
        </div>
        <div
          class="sc-irTrJD OpjW"
        />
        <div
          align="flex-end"
          class="sc-jsMaUZ gtIvBc"
        >
          <div
            class="sc-eDDOjk hzWDLl"
            margin="10px 0 0 0"
          >
            <button
              class="sc-iAEzLF gRRRHm"
              disabled=""
            >
              <div
                class="sc-fsQjcA iKmgdh"
                disabled=""
              >
                &lt;
              </div>
            </button>
            <button
              btnbackgroundcolor="#ffffff"
              class="sc-jTrOVX kwSTpv"
            >
              <div
                class="sc-gLDymS jVaaBL"
                color="#000000"
              >
                1
              </div>
            </button>
            <button
              btnbackgroundcolor="#ffffff"
              class="sc-jTrOVX cLKghP"
            >
              <div
                class="sc-gLDymS TLaLT"
                color="#000000"
              >
                2
              </div>
            </button>
            <button
              class="sc-iAEzLF gRRRHm"
            >
              <div
                class="sc-fsQjcA hFwLtP"
              >
                &gt;
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-qRuZU gmucBJ"
      data-testid="pagination-component"
    >
      <div
        align="space-between"
        class="sc-jsMaUZ dKTrZm"
      >
        <div
          class="sc-kFuvnu iOKxMI"
        >
          <div>
            Show 
          </div>
          <div
            class="sc-beqWNU gbqjiq"
          >
            <select
              class="sc-gueXAH fwRQqY"
            >
              <option
                class="sc-dmqGRC"
                value="5"
              >
                5
              </option>
              <option
                class="sc-dmqGRC"
                value="10"
              >
                10
              </option>
              <option
                class="sc-dmqGRC"
                value="20"
              >
                20
              </option>
            </select>
            <div
              class="sc-hLseSp eewNXK"
            >
              &gt;
            </div>
          </div>
        </div>
        <div
          class="sc-eDDOjk hzWDLl"
          margin="0 0 10px 0"
        >
          <button
            class="sc-iAEzLF gRRRHm"
            disabled=""
          >
            <div
              class="sc-fsQjcA iKmgdh"
              disabled=""
            >
              &lt;
            </div>
          </button>
          <button
            btnbackgroundcolor="#ffffff"
            class="sc-jTrOVX kwSTpv"
          >
            <div
              class="sc-gLDymS jVaaBL"
              color="#000000"
            >
              1
            </div>
          </button>
          <button
            btnbackgroundcolor="#ffffff"
            class="sc-jTrOVX cLKghP"
          >
            <div
              class="sc-gLDymS TLaLT"
              color="#000000"
            >
              2
            </div>
          </button>
          <button
            class="sc-iAEzLF gRRRHm"
          >
            <div
              class="sc-fsQjcA hFwLtP"
            >
              &gt;
            </div>
          </button>
        </div>
      </div>
      <div
        class="sc-irTrJD OpjW"
      />
      <div
        align="flex-end"
        class="sc-jsMaUZ gtIvBc"
      >
        <div
          class="sc-eDDOjk hzWDLl"
          margin="10px 0 0 0"
        >
          <button
            class="sc-iAEzLF gRRRHm"
            disabled=""
          >
            <div
              class="sc-fsQjcA iKmgdh"
              disabled=""
            >
              &lt;
            </div>
          </button>
          <button
            btnbackgroundcolor="#ffffff"
            class="sc-jTrOVX kwSTpv"
          >
            <div
              class="sc-gLDymS jVaaBL"
              color="#000000"
            >
              1
            </div>
          </button>
          <button
            btnbackgroundcolor="#ffffff"
            class="sc-jTrOVX cLKghP"
          >
            <div
              class="sc-gLDymS TLaLT"
              color="#000000"
            >
              2
            </div>
          </button>
          <button
            class="sc-iAEzLF gRRRHm"
          >
            <div
              class="sc-fsQjcA hFwLtP"
            >
              &gt;
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
